File: SkyEditor.SaveEditor.UI.WPF/MysteryDungeon/Explorers/ViewModels/ExplorersItemViewModel.vb
Imports System.ComponentModel
Imports SkyEditor.Core
Imports SkyEditor.Core.UI
Imports SkyEditor.SaveEditor.MysteryDungeon.Explorers

Namespace MysteryDungeon.Explorers.ViewModels
    Public Class ExplorersItemViewModel
        Inherits GenericViewModel(Of ExplorersItem)
        Implements INotifyPropertyChanged
        Implements IClonable

        Public Sub New()
            SetModel(New ExplorersItem())
        End Sub

        Public Sub New(model As Object, appViewModel As ApplicationViewModel)
            MyBase.New(model, appViewModel)
        End Sub

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Public Property ID As Integer
            Get
                Return Model.ID
            End Get
            Set(value As Integer)
                Model.ID = value
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(ID)))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(Name)))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(Quantity)))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(IsBox)))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(IsUsedTM)))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(IsStackableItem)))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(StackVisibility)))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(CanContainItem)))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(ContainedItemVisibility)))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(ContainedItemChoices)))
            End Set
        End Property

        Public ReadOnly Property Name As String
            Get
                Return Model.Name
            End Get
        End Property

        Public Property ContainedItemID As Integer
            Get
                Return Model.ContainedItemID
            End Get
            Set(value As Integer)
                Model.ContainedItemID = value
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(ContainedItemID)))
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(ContainedItemName)))
            End Set
        End Property

        Public ReadOnly Property ContainedItemName As String
            Get
                If TypeOf Model Is TDHeldItem Then
                    Return Lists.TDItems(ContainedItemID)
                Else
                    Return Lists.SkyItems(ContainedItemID)
                End If
            End Get
        End Property

        Public Property Quantity As Integer
            Get
                Return Model.Quantity
            End Get
            Set(value As Integer)
                Model.Quantity = value
                RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(NameOf(Quantity)))
            End Set
        End Property

        Public ReadOnly Property IsBox As Boolean
            Get
                Return Model.IsBox
            End Get
        End Property

        Public ReadOnly Property IsUsedTM As Boolean
            Get
                Return Model.IsUsedTM
            End Get
        End Property

        Public ReadOnly Property IsStackableItem As Boolean
            Get
                Return Model.IsStackableItem
            End Get
        End Property

        Public ReadOnly Property StackVisibility As Visibility
            Get
                If IsStackableItem Then
                    Return Visibility.Visible
                Else
                    Return Visibility.Collapsed
                End If
            End Get
        End Property

        Public ReadOnly Property CanContainItem As Boolean
            Get
                Return IsBox OrElse IsUsedTM
            End Get
        End Property

        Public ReadOnly Property ContainedItemVisibility As Visibility
            Get
                If CanContainItem Then
                    Return Visibility.Visible
                Else
                    Return Visibility.Collapsed
                End If
            End Get
        End Property

        Public ReadOnly Property ContainedItemChoices As Dictionary(Of Integer, String)
            Get
                If TypeOf Model Is TDHeldItem Then
                    If IsBox Then
                        Return Lists.TDItems
                    ElseIf IsUsedTM Then
                        Return Lists.TDItemsMovesOnly
                    Else
                        Return New Dictionary(Of Integer, String)
                    End If
                Else
                    If IsBox Then
                        Return Lists.SkyItems
                    ElseIf IsUsedTM Then
                        Return Lists.SkyItemsMovesOnly
                    Else
                        Return New Dictionary(Of Integer, String)
                    End If
                End If
            End Get
        End Property

        Public Function Clone() As Object Implements IClonable.Clone
            If TypeOf Model Is IClonable Then
                Return New ExplorersItemViewModel(DirectCast(Model, IClonable).Clone(), CurrentApplicationViewModel)
            Else
                Return New ExplorersItemViewModel(Model, CurrentApplicationViewModel)
            End If
        End Function

        Public Overrides Function ToString() As String
            Return If(Name, MyBase.ToString())
        End Function

    End Class
End Namespace

File: SkyEditor.SaveEditor.UI.Avalonia/ViewModels/Explorers/ExplorersItemViewModel.cs
using ReactiveUI;
using SkyEditor.SaveEditor.MysteryDungeon.Explorers;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SkyEditor.SaveEditor.UI.Avalonia.ViewModels.Explorers
{
    public class ExplorersItemViewModel : ViewModelBase
    {
        public ExplorersItemViewModel() : this(new ExplorersItem())
        {
        }

        public ExplorersItemViewModel(ExplorersItem model)
        {
            Model = model ?? throw new ArgumentNullException(nameof(model));
        }

        public ExplorersItem Model { get; }

        public int ID
        {
            get => Model.ID;
            set
            {
                if (Model.ID != value)
                {
                    Model.ID = value;
                    this.RaisePropertyChanged(nameof(ID));
                    this.RaisePropertyChanged(nameof(IDListItem));
                    this.RaisePropertyChanged(nameof(Name));
                    this.RaisePropertyChanged(nameof(Quantity));
                    this.RaisePropertyChanged(nameof(IsBox));
                    this.RaisePropertyChanged(nameof(IsUsedTM));
                    this.RaisePropertyChanged(nameof(IsStackableItem));
                    this.RaisePropertyChanged(nameof(CanContainItem));
                    this.RaisePropertyChanged(nameof(ItemChoices));
                    this.RaisePropertyChanged(nameof(ContainedItemChoices));
                }
            }
        }        

        public ListItem IDListItem
        {
            get => ItemChoices.Count > ID ? ItemChoices[ID] : null;
            set
            {
                if (value != null && Model.ID != value.Value)
                {
                    Model.ID = value.Value;
                    this.RaisePropertyChanged(nameof(ID));
                    this.RaisePropertyChanged(nameof(IDListItem));
                    this.RaisePropertyChanged(nameof(Name));
                    this.RaisePropertyChanged(nameof(Quantity));
                    this.RaisePropertyChanged(nameof(IsBox));
                    this.RaisePropertyChanged(nameof(IsUsedTM));
                    this.RaisePropertyChanged(nameof(IsStackableItem));
                    this.RaisePropertyChanged(nameof(CanContainItem));
                    this.RaisePropertyChanged(nameof(ItemChoices));
                    this.RaisePropertyChanged(nameof(ContainedItemChoices));
                }
            }
        }

        public string Name => Model.Name;

        public int ContainedItemID
        {
            get => Model.ContainedItemID;
            set
            {
                if (Model.ContainedItemID != value)
                {
                    Model.ContainedItemID = value;
                    this.RaisePropertyChanged(nameof(ContainedItemID));
                    this.RaisePropertyChanged(nameof(ContainedItemIDListItem));
                    this.RaisePropertyChanged(nameof(ContainedItemName));
                }
            }
        }

        public ListItem ContainedItemIDListItem
        {
            get => ContainedItemChoices.Count > ContainedItemID ? ContainedItemChoices[ContainedItemID] : null;
            set
            {
                if (value != null && Model.ContainedItemID != value.Value)
                {
                    Model.ContainedItemID = value.Value;
                    this.RaisePropertyChanged(nameof(ContainedItemID));
                    this.RaisePropertyChanged(nameof(ContainedItemIDListItem));
                    this.RaisePropertyChanged(nameof(ContainedItemName));
                }
            }
        }

        public string ContainedItemName => Model.ContainedItemName;

        public int Quantity
        {
            get => Model.Quantity;
            set
            {
                if (Model.Quantity != value)
                {
                    Model.Quantity = value;
                    this.RaisePropertyChanged(nameof(Quantity));
                }
            }
        }

        public bool IsBox => Model.IsBox;

        public bool IsUsedTM => Model.IsUsedTM;

        public bool IsStackableItem => Model.IsStackableItem;

        public bool CanContainItem => IsBox || IsUsedTM;

        public List<ListItem> ItemChoices
        {
            get
            {
                if (Model is TDHeldItem)
                {
                    return Lists.TDItems.Select(kv => new ListItem { DisplayName = kv.Value, Value = kv.Key }).ToList();                    
                }
                else
                {
                    return Lists.SkyItems.Select(kv => new ListItem { DisplayName = kv.Value, Value = kv.Key }).ToList();                 
                }
            }
        }

        public List<ListItem> ContainedItemChoices
        {
            get
            {
                if (Model is TDHeldItem)
                {
                    if (IsBox)
                    {
                        return Lists.TDItems.Select(kv => new ListItem { DisplayName = kv.Value, Value = kv.Key }).ToList();
                    }
                    else if (IsUsedTM)
                    {
                        return Lists.TDItems.Select(kv => new ListItem { DisplayName = kv.Value, Value = kv.Key }).ToList();
                    }
                    else
                    {
                        return new List<ListItem>();
                    }
                }
                else
                {
                    if (IsBox)
                    {
                        return Lists.SkyItems.Select(kv => new ListItem { DisplayName = kv.Value, Value = kv.Key }).ToList();
                    }
                    else if (IsUsedTM)
                    {
                        return Lists.SkyItems.Select(kv => new ListItem { DisplayName = kv.Value, Value = kv.Key }).ToList();
                    }
                    else
                    {
                        return new List<ListItem>();
                    }
                }
            }
        }

        public ExplorersItemViewModel Clone()
        {
            return new ExplorersItemViewModel(Model.Clone());
        }

        public override string ToString()
        {
            return Name ?? base.ToString();
        }
    }
}
